### PR TITLE
Fix range/state/ion/assigned variable ordering

### LINKED ISSUE #436: Wrong order of ion variables declaration in mod file
When compiling `cal_mig.mod` from https://github.com/Neurosim-lab/netpyne/blob/coreneuron/examples/coreneuron/m1/mod/cal_mig.mod the order of the `cai` and `ica` `ion` channels' variables is wrong in the generated `.cpp` file by `NMODL`.
The issue is fixed if `cai` and `ica` definition as `RANGE` variables in `line 44` is removed.
We need to check again the order of the printing of the `ion` variables in `NMODL` to make sure it's the same as `neuron nmodl`.

### PR DESCRIPTION
**Pull request by pramodk**
- range variabes needs to be printed in the order as it's
   done in neuron/mod2c implementation
 - existing implementation had bugs when ion variables are
   state variables. (CaDynamics_E2.mod cal_mig.mod)
 - this PR reviews implementation in mod2c and documents are
   variable order is calculated in original implementation
   and fixes variable ordering issue.

TODO 

- [x] simplify implementation / code
- [x] document some of the implementation details
- [x] add unit tests for variable ordering

fixes #436

### CODE REVIEW

File: src/codegen/codegen_helper_visitor.cpp [hunk 1]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
  *
- * In order to be compatible with NEURON, we need to print range variables in
- * certain order. For example, range variables which are parameters comes first.
- * Also, there is difference between declaration order vs. definition order. For
- * example, POINTER variable in NEURON block is just declaration and doesn't
- * determine the order in which they will get printed. Below we query symbol table
- * and order all instance variables into certain order.
+ * In order to be compatible with NEURON, we need to print all variables in
+ * exact order as NEURON/MOD2C implementation. This is important because memory
+ * for all vatiables is allocated in single 1-D array with certain offset
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:281]:
```suggestion
 * for all variables is allocated in single 1-D array with certain offset
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 2]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ *
+ * We now print the variables in following order:
+ *
+ * - List 1 i.e. range + parameter variables are printed first
+ * - List 3 i.e. range + assigned variables are printed next
+ * - List 2 i.e. range + state variables are printed next
+ * - List 4 i.e. assigned variables not present in previous 3 lists
+ *
+ * NOTE:
+ * - State variables are also has property `assigned_definition` but these variables
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:313]:
```suggestion
 * - State variables also have the property `assigned_definition` but these variables
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 3]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+                                                               // compare by symbol name
+                                                               [&tosearch](
+                                                                   const SymbolType& symbol) {
+                                                                   return tosearch->get_name() ==
+                                                                          symbol->get_name();
+                                                               }) != primary.end();
+                                       }),
+                        secondary.end());
+    };
+
```
Comment (pramodk) [inline src/codegen/codegen_helper_visitor.cpp:361]: @tristan0x : is there any better pattern to achieve above? Or above is "fine"?

File: src/codegen/codegen_helper_visitor.cpp [hunk 4]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+                                                               primary.end(),
+                                                               // compare by symbol name
+                                                               [&tosearch](
+                                                                   const SymbolType& symbol) {
+                                                                   return tosearch->get_name() ==
+                                                                          symbol->get_name();
+                                                               }) != primary.end();
+                                       }),
+                        secondary.end());
+    };
```
Comment (alkino) [inline src/codegen/codegen_helper_visitor.cpp:360]: `auto remote_var_exist = secondary | erase_if([&](const auto& symbol) { return tosearch->get_name() == symbol->get_name();`

Comment (alkino) [reply src/codegen/codegen_helper_visitor.cpp:360]: it's not really the right syntax

Comment (pramodk) [reply src/codegen/codegen_helper_visitor.cpp:360]: interesting! I will see the range-v3 (for future integration)

File: src/codegen/codegen_helper_visitor.cpp [hunk 5]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ * they must be in same order as NEURON.
+ *
+ * Here is how order is determined into NEURON/MOD2C implementation:
+ *
+ * First, following three lists are created
+ * - variables with parameter and range property (List 1)
+ * - variables with state and range property (List 2)
+ * - variables with assigned and range property (List 3)
+ *
+ * Once above three lists are created, we remove following variable from above
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:292]:
```suggestion
 * Once created, we remove following variable from above
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 6]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ * three lists:
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:293]:
```suggestion
 * three lists:
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 7]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ * - In NEURON/MOD2C implementation, we remove variables with NRNPRANGEIN
+ *   or NRNPRANGEOUT type
+ * - So who has NRNPRANGEIN and NRNPRANGEOUT type? these are such USEION read
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:296]:
```suggestion
 * - So who has NRNPRANGEIN and NRNPRANGEOUT type? these are USEION read
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 8]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ * - This is the reason for mod files CaDynamics_E2.mod or cal_mig.mod, ica variable
+ *   is printed earlier in the list but other variables like cai, cao don't appear
+ *   in same order.
+ *
+ * Finally we create 4th list:
+ *  - variables with assigned property and not in the previous 3 lists
+ *
+ * We now print the variables in following order:
+ *
+ * - List 1 i.e. range + parameter variables are printed first
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:307]: nitpick: shouldn't we create the lists in the same order we print them?

File: src/codegen/codegen_helper_visitor.cpp [hunk 9]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ * - So who has NRNPRANGEIN and NRNPRANGEOUT type? these are such USEION read
+ *   or write variables that are not ionic currents.
+ * - This is the reason for mod files CaDynamics_E2.mod or cal_mig.mod, ica variable
+ *   is printed earlier in the list but other variables like cai, cao don't appear
+ *   in same order.
+ *
+ * Finally we create 4th list:
+ *  - variables with assigned property and not in the previous 3 lists
+ *
+ * We now print the variables in following order:
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:305]: nitpick: shouldn't we create the lists in the same order we print them?

File: src/codegen/codegen_helper_visitor.cpp [hunk 10]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ * - List 2 i.e. range + state variables are printed next
+ * - List 4 i.e. assigned and ion variables not present in the previous 3 lists
+ *
+ * NOTE:
+ * - State variables also have the property `assigned_definition` but these variables
+ *   are not from ASSIGNED block.
+ * - Variable can not be range as well as state, it's redeclaration error
+ * - Variable can be parameter as well as range. Without range, parameter
+ *   is considered as global variable i.e. one value for all instances.
+ * - If a variable is only defined as RANGE and not in assgined or parameter
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:318]:
```suggestion
 * - If a variable is only defined as RANGE and not in assigned or parameter
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 11]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ *   or state block then it's not printed.
+ * - Note that if variable property is different than the variable type. For example,
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:320]:
```suggestion
 * - Note that a variable property is different than the variable type. For example,
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 12]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+        vars.erase(std::remove_if(vars.begin(),
+                                  vars.end(),
+                                  [&](SymbolType& s) {
+                                      return info.is_ion_variable(s->get_name()) &&
+                                             !info.is_ionic_current(s->get_name());
+                                  }),
+                   vars.end());
+    };
+
+    /// if `seconday` vector contains any symbol that exist in the `primary` then remove it
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:345]:
```suggestion
    /// if `secondary` vector contains any symbol that exist in the `primary` then remove it
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 13]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+    auto remove_var_exist = [](SymbolVectorType& primary, SymbolVectorType& secondary) -> void {
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:346]: another implementation is to stuff a hashtable with the values of primary and then run over the second and check if they are present in the hash table. This has complexity O(N_p + N_s) on average (N_p = primary.size(), N_s = primary.size()). Yours is O(N_p * N_s). However, hashing has a big constant in front so for small vectors your version could come on top. I would prefer the most O -complexity efficient since in case of small vectors the difference is not noticible anyway. Ultimately your decision

File: src/codegen/codegen_helper_visitor.cpp (hunk 3 above)
Comment (cattabiani) [reply src/codegen/codegen_helper_visitor.cpp:361]: I had a suggestion for better patterns :D

File: src/codegen/codegen_helper_visitor.cpp [hunk 14]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ *   and order all instance variables into certain order.
  */
 void CodegenHelperVisitor::find_range_variables() {
     /// comparator to decide the order based on definition
     auto comparator = [](const SymbolType& first, const SymbolType& second) -> bool {
         return first->get_definition_order() < second->get_definition_order();
     };
 
+    /// from symbols vector `vars`, remove all ion variables which are not ionic currents
+    auto remove_non_ioncur_vars = [](SymbolVectorType& vars, CodegenInfo& info) -> void {
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:335]:
```suggestion
    auto remove_non_ioncur_vars = [](SymbolVectorType& vars, const CodegenInfo& info) -> void {
```

File: src/codegen/codegen_helper_visitor.cpp [hunk 15]
```diff
@@ -615,14 +654,14 @@ void CodegenHelperVisitor::visit_program(const ast::Program& node) {
         }
     }
     node.visit_children(*this);
+    find_ion_variables();
```
Comment (cattabiani) [inline src/codegen/codegen_helper_visitor.cpp:657]: nitpick:
```suggestion
    find_ion_variables(); // Keep this before find_*_range_variables()
```

File: test/unit/codegen/codegen_helper.cpp [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+              cai' = -(10000)*(ica*gamma/(2*FARADAY*depth)) - (cai - minCai)/decay
+            }
+        )";
+
+        THEN("ion state variable is ordered after parameter and assigned ionic current") {
+            std::string expected = "gamma;decay;depth;minCai;ica;cai;";
+            auto result = run_inline_visitor(nmodl_text);
+            REQUIRE(result == expected);
+        }
+    }
```
Comment (cattabiani) [inline test/unit/codegen/codegen_helper.cpp:129]: another edge-case that could be useful to test is `thalamus/mod/ican.mod`. In there there is an ion called `n` and its current could give some problems to the ion current detection system and ispc (since `in` is a reserved variable in there)

**Review by cattabiani — COMMENTED** [review]
Very good! A few nitpicks and a suggestion for improving efficiency

File: test/unit/codegen/codegen_helper.cpp (hunk 16 above)
Comment (pramodk) [reply test/unit/codegen/codegen_helper.cpp:129]: thanks, I will add that shortly.

Comment (pramodk) [reply test/unit/codegen/codegen_helper.cpp:129]: ican.mod is handled normally in the context of this PR : 

image

This mod file has issue due to `in` variable which is keyword in ISPC. This will be handled separately.

File: src/codegen/codegen_helper_visitor.cpp [hunk 17]
```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
...
+ *  - variables with assigned property and not in the previous 3 lists
+ *
+ * We now print the variables in following order:
+ *
+ * - List 1 i.e. range + parameter variables are printed first
+ * - List 3 i.e. range + assigned variables are printed next
+ * - List 2 i.e. range + state variables are printed next
+ * - List 4 i.e. assigned and ion variables not present in the previous 3 lists
+ *
+ * NOTE:
```
Comment (iomaganaris) [inline src/codegen/codegen_helper_visitor.cpp:311]: I suppose that all of the notes below apply to mod2c/neuron as well, right?

File: test/unit/codegen/codegen_helper.cpp [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+    GIVEN("CaDynamics_E2.mod : USEION variables declared as STATE variable") {
+        std::string nmodl_text = R"(
+            NEURON  {
+              SUFFIX CaDynamics_E2
+              USEION ca READ ica WRITE cai
```
Comment (iomaganaris) [inline test/unit/codegen/codegen_helper.cpp:102]: I think that there are some mod files that have the same variable in `READ` and `WRITE`? What happens with those ones? Maybe a test would be useful?
Looking at the code I think that they are handled the same way. The only issue might be that they are declared 2 times?

**Review by iomaganaris — COMMENTED** [review]
LGTM in general.
I have a couple of questions only.

File: test/unit/codegen/codegen_helper.cpp (hunk 18 above)
Comment (pramodk) [reply test/unit/codegen/codegen_helper.cpp:102]: cadyn.mod unit test added. Its handled correctly.

**Review by iomaganaris — APPROVED** [review]

**Comment by cattabiani** [comment]
I am Katta and I approve this PR :D

**Review by cattabiani — APPROVED** [review]
